Fix Lua API polluting the game context

MjnMixael opened the pull request:
I was working on some lua scripts and once again got annoyed by having to save color and font values to restore them when I'm done drawing with Lua. So I whipped up this draft idea as a possible way to solve the issue. This creates `gr_lua_screen` with relevant members. Setting the color, font, or line width from Lua will, by default, only set the members of that struct. Calling drawing methods from Lua, by default, will use the members of that struct. I left in a way to override and set/use the game context values as well but I'm not really certain how useful that is or if it just muddies up the API a bit.

This is all very much draft status and mostly just wanting to get feedback if this is worth cleaning up and implementing in full.

Fixes #4815

MjnMixael commented:
> As for the implementation, you might consider toggling the context via a separate function or variable, rather than having new optional arguments on all these functions.

That's an interesting idea. Do you think there's much need for Lua to mess with the game's internal context? I couldn't think of one but it's hard to know what everyone's done in the past. 

Personally I'd even be ok with disallowing Lua to get or set anything outside of the Lua context.

Baezon commented:
I would be too, but it's hard to think of all the uses cases.  One hand, it was possible before so it'd be a shame to 'take it away', but my general line-in-the-sand for not providing some functionality is when its misuse (or just use) can cause mysterious problems which are difficult to diagnose or workaround on their end, which I'd probably classify this 'polluted context' as. I don't feel particularly strongly one way or the other.

MjnMixael commented:
> I would be too, but it's hard to think of all the uses cases. One hand, it was possible before so it'd be a shame to 'take it away', but my general line-in-the-sand for not providing some functionality is when its misuse (or just use) can cause mysterious problems which are difficult to diagnose or workaround on their end, which I'd probably classify this 'polluted context' as. I don't feel particularly strongly one way or the other.

Ok, so I added a couple bools to the lua context to track what we're doing which simplifies a lot of the parameters passed around. It's not my preferred way but it matches how we use `gr_screen` a bit so I guess that works. I also added a virtvar to set a context override. Tell me what you think of this version.

Baezon commented:
Yep, I'd be happy with this approach.

wookieejedi commented:
Awesome to see this as ready for review, will make scripting in this area so much cleaner.

Baezon commented on code/scripting/api/libs/graphics.cpp:
```diff
@@ -1157,6 +1216,8 @@ ADE_FUNC(drawTargetingBrackets, l_Graphics, "object Object, [boolean draw=true,
 		line_draw_list.flush();
 	}
 
+	gr_lua_screen.active = false;
```
This should come after the `g3_end_frame()` below to match its other usages. I don't think it happens to matter in this case, but it could in the future.

MjnMixael replied:
Good call. I've addressed this and I think everything else.

Baezon reviewed (changes requested):
Ok, took a closer look at it now that it's ready for review.

MjnMixael commented on code/graphics/opengl/gropengl.cpp:
```diff
@@ -787,10 +787,15 @@ void gr_opengl_zbias(int bias)
+		gr_lua_screen.line_width = width;
+	} else {
+		if (width <= 1.0f) {
+			GL_state.SetLineWidth(width);
```
It wasn't before and I'm honestly not sure what the implications of it are so I opted to keep behavior the same.

Baezon replied:
Fair enough.

Baezon reviewed (approved):
Apparently, screen scale and the clip rect have a similar situation going on, did you have any other plans in that regard?

MjnMixael commented:
> I think its fine to merge this and have a follow-up PR. They've got the same sort of set a thing, then reset it situation. You can see in `gr_reset_screen_scale()` the engine maintains a bunch of 'save' variables for this purpose, which isn't the prettiest.

Sounds good. I'll handle that once this is merged.
